### ISSUE TITLE
calc tm breaks with lower case input

### ISSUE THREAD

**Issue by Tim-Kirkwood** [libnano/primer3-py#115]
Hello,
My system: 
```
windows 10
python version 3.11.3
```

I am working in a conda env with these libs:
```
name: pip_env
dependencies:
  - python
  - spyder-kernels=2.4.1
  - biopython
  - plotly
  - pandas
  - pip
  - pip:
    - primer3-py
    - snapgene_reader
```

Feeding in lower case sequences seems to break the calc tm methods:

```
oligo_calc = p3.thermoanalysis.ThermoAnalysis(mv_conc=50, 
                                              dv_conc=1.5,
                                              dntp_conc=0.6, 
                                              dna_conc=50.0)
seq_lower = 'atgcatttcgcatctcgagatctaagaga'
seq_upper = seq_lower.upper()
print ('---OLIGO CALC---\n',
       f'TM UPPER:  {oligo_calc.calc_tm(seq_upper)}\n',
       f'TM LOWER:  {oligo_calc.calc_tm(seq_lower)}\n',
      '---BINDINGS CALC---\n',
      f'TM UPPER:  {p3.bindings.calc_tm(seq_upper)}\n',
      f'TM LOWER:  {p3.bindings.calc_tm(seq_lower)}\n'
       )

# Prints:
# ---OLIGO CALC---
# TM UPPER:  62.01093944856859
# TM LOWER:  -999999.9999
# ---BINDINGS CALC---
# TM UPPER:  62.01093944856859
# TM LOWER:  -999999.9999
```

**Comment by benpruitt** [comment]
The `oligotm` binary that is built w/ the native `primer3` library converts lower case sequences to upper case as a matter of course (see [oligotm_main.c:256](https://github.com/libnano/primer3-py/blob/83018cdacbc8526ebe26902e8a378a231c60e222/primer3/src/libprimer3/oligotm_main.c#L256)). The underlying `oligotm` function, which we invoke directly in the `primer3-py` Cython code, expects a fully upper case input. 

We've generally avoided adding / including input sanitization in `primer3-py` as a matter of course, given that the library is intended to offer a more performant alternative to command line wrappers. I would suggest that you just call `.upper()` on the string input to `calc_tm`. If additional IUPAC degenerate base sanitization is required, `primer3-py` includes a `sanitize_sequence` method in `p3helpers` ([ref](https://github.com/libnano/primer3-py/blob/83018cdacbc8526ebe26902e8a378a231c60e222/primer3/p3helpers.pyx#LL201C5-L201C22))

**Comment by manulera** [comment]
Hi @benpruitt I think this should be handled by the library by either:
* Modifying the sequences to make them uppercase
* Throw an error otherwise.

I am happy to provide a fix for either solution in a PR.

I think it would be way more safe to make sure users don't have to remember or be aware about this unexpected behaviour whenever they use the library (I only realised by chance). The performance difference is probably minimal and it can really help with preventing bugs.

**Comment by manulera** [comment]
cc @peterjc hello here as well!

**Comment by benpruitt** [comment]
I'm on board with throwing an error. It looks like `oligotm` returns that magic `Tm` value (`OLIGOTM_ERROR`; -999999.9999) if the input sequence either contains an invalid character or is too short for the NN calcs (i.e., shorter than 2 bases). 

https://github.com/libnano/primer3-py/blob/83018cdacbc8526ebe26902e8a378a231c60e222/primer3/src/libprimer3/oligotm.c#L557
https://github.com/libnano/primer3-py/blob/83018cdacbc8526ebe26902e8a378a231c60e222/primer3/src/libprimer3/oligotm.h#L46

I'd suggest we do the following:

1. Extend the [tm_ret struct](https://github.com/libnano/primer3-py/blob/83018cdacbc8526ebe26902e8a378a231c60e222/primer3/src/libprimer3/oligotm.h#L48) to include an error field, with a corresponding enum that discriminates between the two error cases
2. Modify [thermoanalysis.pyx](https://github.com/libnano/primer3-py/blob/83018cdacbc8526ebe26902e8a378a231c60e222/primer3/thermoanalysis.pyx#L1012) to check for an error code and conditionally raise a corresponding Python exception 

I'd prefer this to input sanitization/conversion as a matter of course, if you guys are on board.

**Comment by manulera** [comment]
Hi @benpruitt thanks for the quick followup. I was thinking to do the sanitation / error handling at the python level. I am not familiar with C so I would not be able to implement a fix there

**Comment by grinner** [comment]
@benpruitt we would need to modify 
1. `oligodg(...)` 
2. `oligotm(...)` 
3. `long_seq_tm(...)`
4. `seqtm(...)` 

to accommodate this.

When would we do the capitalization check in the C code?

This seems like it makes sense  to do in `thermoanalysis.pyx` but arguments of of type `unsigned char *`
Can we add a `do_capitalization_check: bool` argument there and call it a day?

**Comment by benpruitt** [comment]
Good point @grinner -- part of why I'd suggested we augment the `primer3` library rather than implement an additional check is that `oligotm` natively checks for non-`ATGC` characters during the calculation (so there would be only be `O(1)` cost to adding this, versus `O(n)` with an additional loop-based check). But `long_seq_tm` doesn't actually perform this check because it uses a different calculation method. 

So yes I agree, we can just perform the check in the Cython code (last time we profiled performing the check and/or sanitization in native Python it slowed things down by an order of magnitude, which is unsurprising given CPython loop performance).

**Comment by benpruitt** [comment]
Addressed in the new release (see #159). The behavior is now:
- Lower case sequences comprised solely of `acgt` are handled gracefully (internally capitalizes)
- Sequences with non-`acgtACGT` characters result in a Python exception
